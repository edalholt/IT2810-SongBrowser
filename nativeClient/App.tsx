import { StatusBar } from 'expo-status-bar';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { ApolloClient, InMemoryCache, ApolloProvider } from '@apollo/client';
import { ThemeProvider, Button, createTheme } from '@rneui/themed';
import { useColorScheme } from 'react-native';
import useCachedResources from './hooks/useCachedResources';
import Navigation from './navigation';

export default function App() {
  const isLoadingComplete = useCachedResources();
  const colorScheme = useColorScheme();

  const client = new ApolloClient({
    uri: 'http://it2810-67.idi.ntnu.no:4000/graphql',
    cache: new InMemoryCache()
  });

  const theme = createTheme({
    lightColors: {
      primary: '#52796f',
      white: '#FFD523', 
    },
    darkColors: {
      background: '#2C2E43',
      primary: '#2C2E43',
      white: '#FFD523',
    },
    mode: colorScheme || "dark",
  });
  console.log(theme.mode)
  
  if (!isLoadingComplete) {
    return null;
  } else {
    return (
      <SafeAreaProvider>
        <ThemeProvider theme={theme}>
        <ApolloProvider client={client}>
        <Navigation colorScheme={colorScheme}/>
        <StatusBar />
        </ApolloProvider>
        </ThemeProvider>
      </SafeAreaProvider>
    );
  }
}
