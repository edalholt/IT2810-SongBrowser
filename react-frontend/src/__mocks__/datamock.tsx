import { GET_SONGS } from "../GraphQL/Queries";

export const mocks = [
    {
      request: {
        query: GET_SONGS,
        variables: {
          name: "Sad Forever"
        }
      },
      result: {
        data: {
          songs: { _id:"634b4ee3dad5dd1cb9c722c4", name:"Sad Forever", artists:"Lauv", year:"2020", danceability: "0.527", popularity:"71", duration:"203507"}
        }
      }
    }
  ];