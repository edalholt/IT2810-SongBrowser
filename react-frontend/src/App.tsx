import { useState, useMemo, createContext, useEffect } from "react";
import "./App.css";
import FrontPage from "./components/FrontPage";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import { grey } from "@mui/material/colors";
import { PaletteMode, Button, Typography } from "@mui/material";
import { Brightness4, Brightness7 } from "@mui/icons-material";
import { CssBaseline } from "@mui/material";

export const ColorModeContext = createContext({ switchMode: () => {} });

declare module "@mui/material/styles/createPalette" { //Inspired by: https://stackoverflow.com/questions/60424596/cant-customize-color-palette-types-on-material-ui-theme-in-typescript
  interface Palette {
    custom: { main: string };
    buttonColor: {main: string };
    tableRow: string;
    searchBar: string;
    textColor: string;
  }
  interface PaletteOptions {
    custom: { main: string };
    buttonColor: {main: string };
    tableRow: string;
    searchBar: string;
    textColor: string;
  }
}

declare module '@mui/material/Button' {
  interface ButtonPropsColorOverrides {
    buttonColor: true;
  }
}

declare module'@mui/material/Typography' {
  interface TypographyPropsColorOverrides {
    textColor: true;
  }
}

declare module '@mui/material/TextField' {
  interface TextFieldPropsColorOverrides {
    searchBar: true;
  }
}

const changeTheme = (mode: PaletteMode) => ({
  palette: {
    mode,
    ...(mode === "light"
      ? {
          // Light mode palette
          custom: { 
            main: "#CAD2C5" 
          },
          divider: "#354F52",
          background: {
            default: "#52796f" 
          },
          buttonColor: {
            main: "#CAD2C5"
          },
          tableRow: "white",
          searchBar: "white",
          textColor: "white",
          button:{
            color: '#ffffff', 
            borderColor: '#ffffff'
          }
        }
      : {
          // Dark mode palette
          custom: { 
            main: "#2F3E46"
          },
          divider: grey[200],
          background: {
            default: "#354F52",
          },
          buttonColor: {
            main: "#2F3E46"
          },
          tableRow: grey[900],
          searchBar: grey[900],
          textColor: "white",
        }),
  },
});

function App() {
  const lightMode = localStorage.getItem("lightMode");
  // If stored mode is light use light, else use dark (default)
  const [mode, setMode] = useState<PaletteMode>((lightMode == "light") ? ("light") : ("dark"));
  
  // Save light or dark mode when changed
  useEffect(() => {
    localStorage.setItem("lightMode", String(mode))
  }, [mode])
  
  const colorMode = useMemo(
    () => ({
      switchMode: () => {
        setMode((prevMode: PaletteMode) => 
          prevMode === "light" ? "dark" : "light"
        )
      },
    }),
    []
  );

  const theme = useMemo(() => createTheme(changeTheme(mode)), [mode]);

  return (
    <>
      <Button sx={{m: 1, color: "#fff"}} title="toggle light/dark mode" onClick={colorMode.switchMode}>
        {theme.palette.mode === "dark" ? 
        <><Brightness7/><Typography color={"#fff"} ml={1}>Light Mode</Typography></>
        : <><Brightness4 /><Typography color={"#fff"} ml={1}>Dark Mode</Typography></>}
      </Button>
      <div className="App">
      <ColorModeContext.Provider value={colorMode}>
        <ThemeProvider theme={theme}>
          <CssBaseline />
          <FrontPage />
        </ThemeProvider>
      </ColorModeContext.Provider>
    </div>
    </>
  );
}

export default App;
