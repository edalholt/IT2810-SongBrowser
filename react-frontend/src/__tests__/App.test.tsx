import React from 'react';
import { render} from '@testing-library/react';
import App from '../App';
import {mocks} from '../__mocks__/datamock'; 
import { MockedProvider } from '@apollo/client/testing';

it("should match snapshot", () => {
  render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <App />
    </MockedProvider>
  );
  expect(mocks).toMatchSnapshot();
});


