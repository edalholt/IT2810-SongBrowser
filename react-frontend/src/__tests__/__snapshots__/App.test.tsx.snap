// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Array [
  Object {
    "request": Object {
      "query": Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": undefined,
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "search",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "search",
                        },
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "page",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "page",
                        },
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "pageSize",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "pageSize",
                        },
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "orderBy",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "orderBy",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "getSongs",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "songs",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "_id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "artists",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "danceability",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "year",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "duration_ms",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "explicit",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "popularity",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "rating",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "page",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "totalPages",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "search",
                  },
                },
              },
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "Int",
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "page",
                  },
                },
              },
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "Int",
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "pageSize",
                  },
                },
              },
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "OrderBySelect",
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "orderBy",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 371,
          "start": 0,
        },
      },
      "variables": Object {
        "name": "Sad Forever",
      },
    },
    "result": Object {
      "data": Object {
        "songs": Object {
          "_id": "634b4ee3dad5dd1cb9c722c4",
          "artists": "Lauv",
          "danceability": "0.527",
          "duration": "203507",
          "name": "Sad Forever",
          "popularity": "71",
          "year": "2020",
        },
      },
    },
  },
]
`;
